workers/agent/agent.go

```diff
@@ -0,0 +1,713 @@
+
+	// certFile represents the local path to the agent cert
+	certFile string
+
+	// keyFile represents the local path to the agent key
+	keyFile string
+
+	// caCertFile represents the local path to the agent ca cert
+	caCertFile string
+
+	// Instance of scheduler
+	scheduler scheduler.GaiaScheduler
+
+	// Instance of store
+	store store.GaiaStore
+
+	// Signal channel for this agent
+	sigs chan os.Signal
+}
+
+// InitAgent initiates the agent instance
+func InitAgent(scheduler scheduler.GaiaScheduler, store store.GaiaStore, certPath string) *Agent {
+	ag := &Agent{
+		scheduler: scheduler,
+		store:     store,
+	}
+
+	// Set path to local certificates
+	ag.certFile = filepath.Join(certPath, "cert.pem")
+	ag.keyFile = filepath.Join(certPath, "key.pem")
+	ag.caCertFile = filepath.Join(certPath, "caCert.pem")
+
+	// return instance
+	return ag
+}
+
+// StartAgent starts the agent main loop and waits until SIGINT or SIGTERM
+// signal has been received.
+func (a *Agent) StartAgent() error {
+	// Allocate SIG channel
+	a.sigs = make(chan os.Signal, 1)
+
+	// Register the signal channel
+	signal.Notify(a.sigs, syscall.SIGINT, syscall.SIGTERM)
+
+	// Setup connection information
+	clientTLS, err := a.setupConnectionInfo()
+	if err != nil {
+		return err
+	}
+
+	// Setup gRPC connection
+	dialOption := grpc.WithTransportCredentials(clientTLS)
+	conn, err := grpc.Dial(gaia.Cfg.WorkerGRPCHostURL, dialOption)
+	if err != nil {
+		return fmt.Errorf("failed to connect to remote host: %s", err.Error())
+	}
+	defer conn.Close()
+
+	// Get worker interface
+	a.client = pb.NewWorkerClient(conn)
+
+	// Start periodic go routine which schedules the worker work
+	workTicker := time.NewTicker(schedulerTickerSeconds * time.Second)
+	quitScheduler := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-workTicker.C:
+				// execute schedule function
+				a.scheduleWork()
+			case <-quitScheduler:
+				workTicker.Stop()
+				return
+			}
+		}
+	}()
+
+	// Start periodic go routine which sends back information to the Gaia primary instance
+	updateTicker := time.NewTicker(updateTickerSeconds * time.Second)
+	quitUpdate := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-updateTicker.C:
+				// run update function
+				a.updateWork()
+			case <-quitUpdate:
+				updateTicker.Stop()
+				return
+			}
+		}
+	}()
+
+	// Block until signal received
+	<-a.sigs
+	gaia.Cfg.Logger.Info("exit signal received. Exiting...")
+
+	// Safely stop scheduler
+	close(quitScheduler)
+	close(quitUpdate)
+
+	return nil
+}
+
+// setupConnectionInfo setups the connection info object by parsing existing
+// mTLS certificates or registering the worker at the Gaia primary instance
+// and receiving new mTLS certs.
+func (a *Agent) setupConnectionInfo() (credentials.TransportCredentials, error) {
+	// Evaluate and summarize all worker tags
+	var tags []string
+	if gaia.Cfg.WorkerTags != "" {
+		trimmedTags := strings.ReplaceAll(gaia.Cfg.WorkerTags, " ", "")
```

Comment:
@michelvocks hm, this means that we aren't allowing for tags with spaces in them. That is okay, I guess, right?

Reply:
Correct. Tags are single words which can be combined with dashes (for example). From my experience, that is the expected behavior.